### ISSUE TITLE
unknown error -505

### ISSUE THREAD

**Issue by messeant** [dmfs/tasks#121]
Hi,

I have an unknown error 505 when I want to install the application (on Nexus 4 Android 5.0.1)

Thanks

**Comment by dmfs** [comment]
Did you install any other app that may access the tasks? Error -505 occurs when another app defines the same permissions. Since Android 5 this is only allowed when the other app is signed by the same key. I hope they change that behavior.

**Comment by messeant** [comment]
I have install CalDAV-Sync, and i must install Tasks.

**Comment by dmfs** [comment]
Did you compile the task app yourself?

**Comment by messeant** [comment]
I haven't an IDE with Android SDK on my computer

**Comment by messeant** [comment]
Even  if it can work in compiling the application, I think that should be corrected so that it works from Google Play

**Comment by dmfs** [comment]
Actually the Google Play version should always work. You'll may into problems when you compile it yourself. The Google Play Versions of our CalDAV-Sync app and of the Task app are signed by the same key.

Can you try to get a system log? The log should show what the actual problem is.

**Comment by messeant** [comment]
where is the system log on Android?

**Comment by dmfs** [comment]
If you have root access you can use an app like CatLog. If you have the Android SDK you can use `adb logcat`. Unfortunatley there is no easy way to get the log without root access.

**Comment by messeant** [comment]
I haven't a rot access.

**Comment by messeant** [comment]
I have installed a old version (1.0.12) from amazon app shop

**Comment by dmfs** [comment]
All our versions are available here: https://github.com/dmfs/tasks/tree/master/releases
I won't be able to tell what exactly the problem is without a log, sorry. Maybe you can try to install adb on your computer and capture the log. That works without root. You just need to enable developer options and USB debugging.

**Comment by messeant** [comment]
I can't install from tasks-1.1.2.apk.

I try to install android SDK to use adb

**Comment by dmfs** [comment]
Interesting. Maybe Error -505 has other reasons as well. Thanks for helping to debug this.

**Comment by messeant** [comment]
I can't connect my Nexus with adb :-( 
But i have generet a bug report file ( contains the  SYSTEM LOG (logcat -v threadtime -d *:v) )

This is an extract from the SYSTEM LOG when I have the error -505:

02-17 16:10:09.646  1540  1540 D Finsky  : [1] InstallerTask.startInstallingGzippedApk: Prepare to unzip org.dmfs.tasks from content://downloads/my_downloads/4510
02-17 16:10:09.848  1540  1540 D Finsky  : [1] 9.onPostExecute: Successfully unpacked gzipped APK to update org.dmfs.tasks
02-17 16:10:10.158  1540  1540 D Finsky  : [1] 1.onPostExecute: Committing session 487864670 for org.dmfs.tasks
02-17 16:10:10.158  1540  1540 D Finsky  : [1] 4.installBeginning: Begin install of org.dmfs.tasks
02-17 16:10:10.639   556   590 I ActivityManager: Start proc com.android.defcontainer for service com.android.defcontainer/.DefaultContainerService: pid=9920 uid=10013 gids={50013, 9997, 1028, 1015, 1023, 2001} abi=armeabi-v7a
02-17 16:10:10.776  1540  1540 D Finsky  : [1] PackageVerificationReceiver.onReceive: Verification requested, id = 0
02-17 16:10:10.782  1540  1540 D Finsky  : [1] WorkerTask.onPreExecute: Verification Requested for id = 0, data=file:///data/app/vmdl487864670.tmp flags=18 fromVerificationActivity=false
02-17 16:10:10.785  1540  1667 D Finsky  : [160] PackageVerificationService.getPackageInfo: Cannot read archive for file:///data/app/vmdl487864670.tmp in request id=0
02-17 16:10:10.788   556   590 D PackageManager: /data/app/vmdl487864670.tmp already staged; skipping copy
02-17 16:10:10.792  1540  1540 D Finsky  : [1] PackageVerificationReceiver.onReceive: Verification requested, id = 0
02-17 16:10:11.227   556   590 W PackageManager: Package org.dmfs.tasks attempting to redeclare permission org.dmfs.permission.WRITE_TASKS already owned by org.dmfs.caldav.lib
02-17 16:10:11.239   556   724 I ActivityManager: Killing 3764:com.google.android.calendar/u0a6 (adj 15): empty #17
02-17 16:10:11.352   556   773 W libprocessgroup: failed to open /acct/uid_10006/pid_3764/cgroup.procs: No such file or directory
02-17 16:10:11.414   556   590 I art     : Explicit concurrent mark sweep GC freed 148801(6MB) AllocSpace objects, 35(1330KB) LOS objects, 33% free, 26MB/40MB, paused 1.648ms total 182.084ms
02-17 16:10:11.425  1540  1540 D Finsky  : [1] PackageInstallerImpl.cancelSession: Canceling session 487864670 for org.dmfs.tasks
02-17 16:10:11.425  1540  1540 E Finsky  : [1] PackageInstallerImpl.handleCommitCallback: Error -505 while installing org.dmfs.tasks: INSTALL_FAILED_DUPLICATE_PERMISSION: Package org.dmfs.tasks attempting to redeclare permission org.dmfs.permission.WRITE_TASKS already owned by org.dmfs.caldav.lib
02-17 16:10:11.426  1540  1540 W Finsky  : [1] 4.installFailed: Install failure of org.dmfs.tasks: -505 null
02-17 16:10:11.446  1540  1540 D Finsky  : [1] InstallerTask.cancelCleanup: Cancel running installation of org.dmfs.tasks
02-17 16:10:11.457  1540  1540 D Finsky  : [1] 1.releasePackage: User=UserHandle{0} released=org.dmfs.tasks

**Comment by messeant** [comment]
perghaps : PackageManager: Package org.dmfs.tasks attempting to redeclare permission org.dmfs.permission.WRITE_TASKS already owned by org.dmfs.caldav.lib

**Comment by dmfs** [comment]
Thanks! That actually should work. I'll have to ask on the Android developer mailing list. It works for most users. That might be an Android bug. Did you install CalDAV-Sync via Google Play?

**Comment by messeant** [comment]
No, I have buy caldav-Sync on Amazon app Shop.

**Comment by messeant** [comment]
it is not Ok. with tasks 1.1.2 on google play.

**Comment by dmfs** [comment]
I see, but the version form Amazon should work just as well. I'll update the task app at Amazon soon. Maybe you try again then.

**Comment by messeant** [comment]
Is it because of Android 5 I have this problem.?

**Comment by dmfs** [comment]
Well, the strict check has been introduced in Android 5, but nevertheless should it work.

**Comment by messeant** [comment]
this error saying it work not:

02-17 16:10:11.425 1540 1540 E Finsky : [1] PackageInstallerImpl.handleCommitCallback: Error -505 while installing org.dmfs.tasks: INSTALL_FAILED_DUPLICATE_PERMISSION: Package org.dmfs.tasks attempting to redeclare permission org.dmfs.permission.WRITE_TASKS already owned by org.dmfs.caldav.lib

There is a conflict between caldav_sync and tasks .

**Comment by dmfs** [comment]
I know. But these versions work for thousands of users. The root of this evil is a general issue with Android permissions. A quick explanation:

When you install an app you agree to the permissions that the store presents you. The permissions an app needs are hard coded into an XML file.
Of course Android needs a definition of each permission that an app requests, otherwise it can't present that permission to the user (because the definition contains a human readable description).
If it can't present the permission to the user (because there is no permission definition), the user can not agree to the permission.
if the user can not agree, the permission is not granted (that's what you would expect).

The problem:

In our case the user in general installs CalDAV-Sync first and then the task app. If the task app (which defines the permission to read and write tasks) wasn't installed when CalDAV-Sync was installed the read/write task permissions are not granted to CalDAV-Sync, so it can't write tasks. You will have to uninstall and reinstall CalDAV-Sync to make this work. Which is very annoying, because you have to set up your account again.

The solution was to add the definition of the permissions to CalDAV-Sync as well. That way the permissions were known and granted when the user installed CalDAV-Sync. This worked fine until Android 5. Now it requires that apps that define the same permissions are signed by the same key. That's already annoyingly enough, but it seems to work for most of our users.

However, it looks like in some cases the same signature is not enough anymore.

So the root of this issue is permission management in Android. If Android would have an on-demand permission management like iOS we wouldn't have this issue, because there wouldn't be a reason do redefine the permissions in CalDAV-Sync.

The other "solution" would be to drop the permissions to read/write tasks, which is probably not a good idea, because it would allow any app to change tasks without knowledge of the user.

**Comment by scintill** [comment]
I had this issue with trying to install OpenTasks 1.1.8.2 from F-Droid with CalDav-Sync 0.4.16 from Play Store (I stopped using Google apps, and Google also locked me out of that account for changing IP addresses, is why it's an old version.) The order I installed them didn't seem to help.

I was able to install https://github.com/dmfs/opentasks/blob/master/releases/opentasks-1.1.8.1.apk .  I am guessing that the F-Droid version has a different key.

**Comment by dmfs** [comment]
@scintill Yep, that's exactly the problem. We'll keep on publishing releases signed with our key in our github repo and we're planning to provide an F-Droid repo so you can get updates more easily.

That reminds me to push the 1.1.8.2 build :-|
